qemu/qmp.go

```diff
@@ -672,6 +672,44 @@ func (q *QMP) ExecuteDeviceAdd(ctx context.Context, blockdevID, devID, driver, b
 	return q.executeCommand(ctx, "device_add", args, nil)
 }
 
+// ExecuteSCSIDeviceAdd adds the guest portion of a block device to a QEMU instance
+// using a SCSI driver with the device_add command.  blockdevID should match the
+// blockdevID passed to a previous call to ExecuteBlockdevAdd.  devID is the id of
+// the device to add.  Both strings must be valid QMP identifiers.  driver is the name of the
+// scsi driver,e.g., scsi-hd, and bus is the name of a SCSI controller bus.
+// scsiID is the SCSI id, lun is logical unit number. scsiID and lun are optional.
+func (q *QMP) ExecuteSCSIDeviceAdd(ctx context.Context, blockdevID, devID, driver, bus string, scsiID, lun *int) error {
```

Comment:
```scsiID, lun *int``` seems weird, I think both should be ```string```

Reply:
@devimc These are supposed to be integer values and not strings. These are optional, so I am accepting a int pointer so that an empty(nil) value can be accepted.

Reply:
Is 0 a valid value for these parameters or is there a known invalid value?  If there were we could avoid using pointers.  Pointers aren't great as they make the API more complicated to use for the end user, as they force you to define variables for the optional parameters even if you just want to use a constant, e.g., I can't do

ExecuteSCSIDeviceAdd(ctx, blockdevID, devID, driver, bus, &1, &2)

They also increase the possibility of race conditions.  The caller needs to think carefully about calling the function in a go-routine.

What combinations of scsiID and lun are valid.  If one is defined, does the other also need to be defined?  If so, you could define a structure containing both parameters and pass a pointer to the structure, e.g.,

ExecuteSCSIDeviceAdd(ctx, blockdevID, devID, driver, bus, nil)
ExecuteSCSIDeviceAdd(ctx, blockdevID, devID, driver, bus, &SCSIOpts{SCSIID: 1, LUN:2})

You're still passing a pointer so we haven't completely solved the race issue but as we'd be mostly passing pointers to temporary variables it would be easier to avoid.

Reply:
@markdryan Positive values of scsiID and lun are valid including zero. If one is defined, the other need not be defined. I hadnt thought of pointer usage from the race condition point of view. If we want to avoid pointers altogether, I could check for a negative value for scsiID and lun. The user can simply pass say -1 for these.

Reply:
> The user can simply pass say -1 for these.

Great.  That will work nicely.